mountpoint-s3-client/src/imds_crt_client.rs

```diff
@@ -0,0 +1,109 @@
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+use futures::channel::oneshot;
+use futures::Future;
+use mountpoint_s3_crt::auth::imds_client::ImdsClient;
+use mountpoint_s3_crt::auth::imds_client::ImdsClientConfig;
+use mountpoint_s3_crt::common::allocator::Allocator;
+use mountpoint_s3_crt::common::error;
+use mountpoint_s3_crt::io::channel_bootstrap::{ClientBootstrap, ClientBootstrapOptions};
+use mountpoint_s3_crt::io::event_loop::EventLoopGroup;
+use mountpoint_s3_crt::io::host_resolver::{HostResolver, HostResolverDefaultOptions};
+use mountpoint_s3_crt::io::retry_strategy::{RetryStrategy, StandardRetryOptions};
+use pin_project::pin_project;
+use thiserror::Error;
+
+#[derive(Debug)]
+/// Instance Metadata Service (IMDS) client responsible for sending ec2 instance metadata query requests.
+pub struct ImdsCrtClient {
+    imds_client: ImdsClient,
+    event_loop_group: EventLoopGroup,
+    allocator: Allocator,
+}
+
+impl ImdsCrtClient {
+    pub fn new() -> Result<Self, error::Error> {
+        let allocator = Allocator::default();
+
+        let mut event_loop_group = EventLoopGroup::new_default(&allocator, None, || {}).unwrap();
+
+        let resolver_options = HostResolverDefaultOptions {
+            max_entries: 8,
+            event_loop_group: &mut event_loop_group,
+        };
+
+        let mut host_resolver = HostResolver::new_default(&allocator, &resolver_options).unwrap();
+
+        let bootstrap_options = ClientBootstrapOptions {
+            event_loop_group: &mut event_loop_group,
+            host_resolver: &mut host_resolver,
+        };
+
+        let client_bootstrap = ClientBootstrap::new(&allocator, &bootstrap_options).unwrap();
+
+        let mut retry_strategy_options = StandardRetryOptions::default(&mut event_loop_group);
+        let retry_strategy = RetryStrategy::standard(&allocator, &retry_strategy_options).unwrap();
+
+        let mut client_config = ImdsClientConfig::new();
+
+        client_config
+            .client_bootstrap(client_bootstrap)
+            .retry_strategy(retry_strategy);
+
+        let imds_client = ImdsClient::new(&allocator, client_config).unwrap();
+
+        Ok(Self {
+            imds_client,
+            event_loop_group,
+            allocator,
+        })
+    }
+
+    /// Query the instnace type mountpoint-s3 running on.
+    pub fn make_instance_type_query(
+        &self,
+        callback: impl FnOnce(Result<String, error::Error>) -> Result<Result<String, error::Error>, ImdsQueryRequestError>
+            + 'static,
+    ) -> Result<ImdsQueryRequest, ImdsQueryRequestError> {
+        let (tx, rx) = oneshot::channel::<Result<Result<String, error::Error>, ImdsQueryRequestError>>();
+        let _ = self.imds_client.get_instance_type(move |result| {
+            let res = callback(result);
+            let _ = tx.send(res);
+        });
+
+        Ok(ImdsQueryRequest { receiver: rx })
+    }
```

Comment:
Why does this method need a callback? The returned `ImdsQueryRequest` is a `Future` that resolves to the returned `String`, or just an error. So anything the callback could do can instead just be done by the caller.

Also, nested `Result` types are a hint that we haven't quite got the interface right.

So I think this method ends up being something like:
```suggestion
    pub fn make_instance_type_query(&self) -> Result<ImdsQueryRequest, ImdsQueryRequestError> {
        let (tx, rx) = oneshot::channel();
        let _ = self.imds_client.get_instance_type(move |result| {
            let _ = tx.send(result.map_err(ImdsQueryRequestError::CrtError));
        });

        Ok(ImdsQueryRequest { receiver: rx })
    }
```

Reply:
fixed

Reply:
Oh, except we don't want to ignore the synchronous error case:
```suggestion
    pub fn make_instance_type_query(&self) -> Result<ImdsQueryRequest, ImdsQueryRequestError> {
        let (tx, rx) = oneshot::channel();
        self.imds_client.get_instance_type(move |result| {
            let _ = tx.send(result.map_err(ImdsQueryRequestError::CrtError));
        })?;

        Ok(ImdsQueryRequest { receiver: rx })
    }
```

Reply:
yeah, my mental model is there are two errors might happen: 
1. Error happened in CRT which is `moutpoint_s3::common::error::Error`
2. Error happened during the data transfer in `oneshot` channel.